server/db/postgres/adapter.go

```diff
@@ -0,0 +1,2695 @@
+		}
+		return true, err
+	}
+	return true, tx.Commit()
+}
+
+// CredIsConfirmed returns true of the given validation method is confirmed.
+func (a *adapter) CredIsConfirmed(uid t.Uid, method string) (bool, error) {
+	var done int
+	// There could be more than one credential of the same method. We just need one.
+	err := a.db.Get(&done, "SELECT done FROM credentials WHERE userid=? AND method=? AND done=true",
+		store.DecodeUid(uid), method)
+	if err == sql.ErrNoRows {
+		// Nothing found, clear the error, otherwise it will be reported as internal error.
+		err = nil
+	}
+
+	return done > 0, err
+}
+
+// credDel deletes given validation method or all methods of the given user.
+// 1. If user is being deleted, hard-delete all records (method == "")
+// 2. If one value is being deleted:
+// 2.1 Delete it if it's valiated or if there were no attempts at validation
+// (otherwise it could be used to circumvent the limit on validation attempts).
+// 2.2 In that case mark it as soft-deleted.
+func credDel(tx *sqlx.Tx, uid t.Uid, method, value string) error {
+	constraints := " WHERE userid=?"
+	args := []interface{}{store.DecodeUid(uid)}
+
+	if method != "" {
+		constraints += " AND method=?"
+		args = append(args, method)
+
+		if value != "" {
+			constraints += " AND value=?"
+			args = append(args, value)
+		}
+	}
+
+	if method == "" {
+		_, err := tx.Exec("DELETE FROM credentials"+constraints, args...)
+		return err
+	}
+
+	// Case 2.1
+	if _, err := tx.Exec("DELETE FROM credentials"+constraints+" AND (done=true OR retries=0)", args...); err != nil {
+		return err
+	}
+
+	// Case 2.2
+	args = append([]interface{}{t.TimeNow()}, args...)
+	_, err := tx.Exec("UPDATE credentials SET deletedat=?"+constraints, args...)
+
+	return err
+}
+
+// CredDel deletes either credentials of the given user. If method is blank all
+// credentials are removed. If value is blank all credentials of the given the
+// method are removed.
+func (a *adapter) CredDel(uid t.Uid, method, value string) error {
+	tx, err := a.db.Beginx()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			tx.Rollback()
+		}
+	}()
+
+	err = credDel(tx, uid, method, value)
+	if err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// CredConfirm marks given credential method as confirmed.
+func (a *adapter) CredConfirm(uid t.Uid, method string) error {
+	res, err := a.db.Exec(
+		"UPDATE credentials SET updatedat=?,done=true,synthetic=CONCAT(method,':',value) "+
+			"WHERE userid=? AND method=? AND deletedat IS NULL AND done=false",
+		t.TimeNow(), store.DecodeUid(uid), method)
+	if err != nil {
+		if isDupe(err) {
+			return t.ErrDuplicate
+		}
+		return err
+	}
+	if numrows, _ := res.RowsAffected(); numrows < 1 {
+		return t.ErrNotFound
+	}
+	return nil
+}
+
+// CredFail increments failure count of the given validation method.
+func (a *adapter) CredFail(uid t.Uid, method string) error {
+	_, err := a.db.Exec("UPDATE credentials SET updatedat=?
```

Comment:
Method signature has changed.

Reply:
Github is buggy. It attaches comments to wrong lines. This comments is about 
`store.RegisterAdapter(adapterName, &adapter{})`